### PR TITLE
Add vectorize support for C++ backend, targets clang and gcc.

### PR DESCRIPTION
**Pull request by zvookin**
Turns out gcc and clang have vector support that makes this mostly doable.

Open to suggestions on testing.

Main use case is generating Apple bitcode via the C++ backend and compiling with Xcode.

### CODE REVIEW

**Comment by zvookin** [comment]
Will fix https://github.com/halide/Halide/issues/1365

**Comment by jrk** [comment]
Do we know whether it's actually useful to have bitcode that doesn't correspond exactly to the intended machine binary? (i.e., is it reasonable to generate bitcode from C while making a binary through our usual custom path?)

**Comment by zvookin** [comment]
Can you elaborate a bit on what you are thinking of when you say
"reasonable"? More specifically are you asking whether Apple's tool chain
requires the native object code be generated from the bitcode that is
embedded? It is Apple so there are no docs on how to write a third party
compiler. My best guess is that Apple is intending to do server side
bitcode to device-native for ARM variants and not much else. At the very
least the address width, endianness, and struct layout can't change.

If there is a stronger requirement, it likely means the only option for
Halide users is to generate only C++ code via the C++ backend and then
compile it to all the variants with Xcode. When we get vectorization,
threading, and GPU execution support for the C++ backend, this is probably
even fairly viable.

If Apple is placing strict requirements on the correspondence between the
object code and the bitcode, the only alternative I see would be for us to
generate the bitcode container as well. We could do this but there are no
official docs on the container format and if there are correspondence
requirements, there are probably other requirements too. I was hoping this
would be a "reasonable" compromise to meet the requirement so folks can use
Halide and ship code on Apple's platforms.

-Z-

**Comment by zvookin** [comment]
I suppose I should also point out that I'm pretty sure the final executable
is always all from bitcode or all from native code. I don't think they can
ever mix and match the types of compiler output. Largely going from the
requirement that if one is compiling to bitcode, all object code going into
the link phase must have bitcode enabled.

-Z-

File: src/CodeGen_OpenCL_Dev.cpp [hunk 1]
```diff
@@ -126,6 +130,7 @@ void CodeGen_OpenCL_Dev::CodeGen_OpenCL_C::visit(const For *loop) {
     }
 }
 
+#if 0
```
Comment (steven-johnson) [inline src/CodeGen_OpenCL_Dev.cpp]: dangling #if 0 here?

File: src/CodeGen_C.cpp [hunk 2]
```diff
@@ -106,12 +107,34 @@ const string globals =
     "template<typename A, typename B> A reinterpret(B b) {A a; memcpy(&a, &b, sizeof(a)); return a;}\n"
     "inline float float_from_bits(uint32_t bits) {return reinterpret<float, uint32_t>(bits);}\n"
     "\n"
-    "template<typename T> T max(T a, T b) {if (a > b) return a; return b;}\n"
-    "template<typename T> T min(T a, T b) {if (a < b) return a; return b;}\n"
+    "template<typename T> T max(T a, T b) {return (a > b) ? a : b;}\n"
```
Comment (steven-johnson) [inline src/CodeGen_C.cpp]: Does our generated C assume/require C++11 yet? If not, should it?

**Comment by jrk** [comment]
Sorry, by "reasonable" I meant "not likely to break anything or do anything too unexpected?"

The most likely unexpected outcome I was imagining is that the bitcode supersedes the native code on many real targets of interest, and that this tanks performance for the deployed app relative to the locally built copy.

**Comment by jrk** [comment]
…none of which is to say that C-backend vectorization support is not a fantastic thing to add!

I suppose the strongest resolution to the above concern would be contacting Apple developer support directly, or bugging someone at WWDC, to lay out our situation and ask what is likely to happen. Perhaps one of the app developer users of this could do that, since they've been the ones asking for an Apple bitcode path?

File: src/CodeGen_C.cpp [hunk 3]
```diff
@@ -1447,7 +1680,63 @@ void CodeGen_C::visit(const Evaluate *op) {
...
+       // TODO: Should this be -1?
+        stream << prefix << 0;
+        prefix = ", ";
+    }
+    stream << "};\n";
+
+    if (op->type != op->vectors[0].type()) {
+        // Union punning only works on little endian machines, but Halide
+        // generally has little-endian deps and it may be faster.
+#if 0
```
Comment (steven-johnson) [inline src/CodeGen_C.cpp]: Another dangling if 0

**Comment by steven-johnson** [comment]
LGTM.

**Comment by steven-johnson** [comment]
Looks like this is generating the vector typedefs into the .h files (even for cases were we aren't generating C++ source) -- pretty sure we don't want that.

File: src/CodeGen_C.cpp [hunk 4]
```diff
@@ -471,6 +572,7 @@ class ExternCallPrototypes : public IRGraphVisitor {
 }
 
 void CodeGen_C::compile(const Module &input) {
+    add_vector_typedefs(input);
```
Comment (steven-johnson) [inline src/CodeGen_C.cpp]: I think this line needs to be

```
    if (!is_header()) {
        add_vector_typedefs(input);
    }
```

**Comment by steven-johnson** [comment]
Note to reviewers: I'd like to land this soon, to unblock some people upstream; assuming the presubmits pass now, the only thing I think it needs is more testing, which I'm going to try to add today. Please comment if you disagree.

**Comment by steven-johnson** [comment]
FYI: the C backend doesn't emit the _argv wrapper or the _metadata() accessor; this isn't a deal breaker for now but should be noted. I'll open an issue.

**Comment by steven-johnson** [comment]
I've added Make rules to build many of the test/generators with this, which has turned up a few codegen bug; I'm investigating those before pushing the expanded testing.

**Comment by steven-johnson** [comment]
I split the needs-additional-tests code into PR#2083; that should probably land first, after which it will be easier to add relevant tests here.

**Comment by steven-johnson** [comment]
Looks like GCC only supports vectors with power-of-2 widths, and has no plans to improve this.

File: src/IROperator.h [hunk 5]
```diff
@@ -188,6 +188,10 @@ struct BufferBuilder {
     EXPORT Expr build() const;
 };
 
+/** If e is a ramp expression with stride, default 1, return the base,
+ * otherwise undefined. */
+Expr strided_ramp_base(Expr e, int stride = 1);
```
Comment (dsharletg) [inline src/IROperator.h]: The code that calls this is pretty confusing when it uses the default argument of 1. I couldn't figure out what was going on until I found this declaration.

Comment (steven-johnson) [inline src/IROperator.h]: Even more confusing, I don't see that any call site uses anything except the default. @zvookin, was there an intended use for this that went missing, or should we just simplify it away?

**Comment by kgnk** [comment]
IMHO, depending on the target, `stride = 2` or even `stride = 4` might be best implemented as one or more vector loads and then shuffles to select the proper elements.

**Comment by abadams** [comment]
CodeGen_LLVM does indeed compile things that way, but we still have ramps
with strides of not one in the IR during lowering.

**Comment by steven-johnson** [comment]
This is a major refactoring of the C++ backend; the idea is that we provide two templated wrapper classes for vectors: one that is pure C++, and one that layers on top of the Clang/C++ native-vector support. They support the same APIs and are selected via a typedef at compile time. Notes:

-- This is deliberately incomplete; in particular, the Shuffle and Select operators need further attention; however, since currently-enabled tests don't exercise these ops at all, the throwing of an internal error is acceptable for now.
-- We deliberately wrap the native-vector stuff to ensure that both C++ and native provide the same API surface.
-- The vector classes are intended to be just enough to support what we need, rather than be-all-end-all vector classes. (e.g. we never emit +=, so we don't bother implementing it.)

I think this is a good approach going forward for robust vector support, so I'm going to continue to flesh it out; I'm deliberately pushing out this incomplete approach to get feedback.

**Comment by zvookin** [comment]
Re: "They support the same APIs and are selected via a typedef at compile time." this should probably explicitly say "final C++ compile time." (There can be at least three compilations: C++ of Halide source, Halide compile, compilation of generated C++.)

**Comment by steven-johnson** [comment]
Pretty sure latest commit also addresses #2078

**Comment by steven-johnson** [comment]
Apparently MSVC 2015 has a string-literal-size-limit of ~16k bytes.

**Comment by steven-johnson** [comment]
OK, assuming this passes the buildbots, I think it's clean enough to go ahead and land: there are still some outstanding issues with the C++ backend, but I think this is a stable enough spot that it's good to move forward with. PTAL.

**Comment by steven-johnson** [comment]
Finally clean. PTAL.

**Comment by steven-johnson** [comment]
Gentle review ping.

File: src/CodeGen_C.cpp [hunk 6]
```diff
@@ -353,6 +395,777 @@ string type_to_c_type(Type type, bool include_space, bool c_plus_plus = true) {
...
+
+#if !defined(__has_builtin)
+    #define __has_builtin(x) 0
+#endif 
+
+template <typename ElementType_, size_t Lanes_>
+class CppVector {
+public:
+    typedef ElementType_ ElementType;
+    static const size_t Lanes = Lanes_;
```
Comment (abadams) [inline src/CodeGen_C.cpp]: What's this for?

**Comment by abadams** [comment]
LGTM, just had that one question

File: src/CodeGen_C.cpp (hunk 6 above)
Comment (steven-johnson) [inline src/CodeGen_C.cpp]: Exposing the arguments to the template as internal-but-public typedefs allows other pieces of code to inspect the particular vector variant in question; the convert_from() implementation uses this to sanity-check the "OtherVec" type. (There were formerly a few other uses too, in code that didn't make the final cut.)

**Pull request merged**